### PR TITLE
Fix up verilator Makefile

### PR DESCRIPTION
**Pull request by markmelvin**

### CODE REVIEW

**Comment by themperek** [comment]
@wallento

File: cocotb/share/makefiles/simulators/Makefile.verilator [hunk 1]
```diff
@@ -59,8 +59,8 @@ $(COCOTB_RESULTS_FILE): $(SIM_BUILD)/$(TOPLEVEL) $(CUSTOM_SIM_DEPS) $(COCOTB_LIB
 	-@rm -f $(COCOTB_RESULTS_FILE)
 
 	PYTHONPATH=$(LIB_DIR):$(PWD):$(NEW_PYTHONPATH) LD_LIBRARY_PATH=$(LIB_DIR) $(LIB_LOAD) MODULE=$(MODULE) \
-        TESTCASE=$(TESTCASE) TOPLEVEL=$(TOPLEVEL) TOPLEVEL_LANG=$(TOPLEVEL_LANG) COCOTB_SIM=1 \
-        $< $(PLUSARGS)
+	TESTCASE=$(TESTCASE) TOPLEVEL=$(TOPLEVEL) TOPLEVEL_LANG=$(TOPLEVEL_LANG) COCOTB_SIM=1 \
+	$(SIM_BUILD)/$(TOPLEVEL) $(PLUSARGS)
```
Comment (imphil) [inline cocotb/share/makefiles/simulators/Makefile.verilator]: - These two lines are part of the line starting with PYTHONPATH, can you indent them with one more tab?
- What's the motivation for switching away from `$<`? It should be identical?

File: cocotb/share/makefiles/simulators/Makefile.verilator [hunk 2]
```diff
@@ -69,14 +69,15 @@ debug: $(SIM_BUILD)/$(TOPLEVEL) $(CUSTOM_SIM_DEPS) $(COCOTB_LIBS) $(COCOTB_VPI_L
 	-@rm -f $(COCOTB_RESULTS_FILE)
 
 	PYTHONPATH=$(LIB_DIR):$(PWD):$(NEW_PYTHONPATH) LD_LIBRARY_PATH=$(LIB_DIR) $(LIB_LOAD) MODULE=$(MODULE) \
-        TESTCASE=$(TESTCASE) TOPLEVEL=$(TOPLEVEL) TOPLEVEL_LANG=$(TOPLEVEL_LANG) COCOTB_SIM=1 \
-        gdb --args $< $(PLUSARGS)
+	TESTCASE=$(TESTCASE) TOPLEVEL=$(TOPLEVEL) TOPLEVEL_LANG=$(TOPLEVEL_LANG) COCOTB_SIM=1 \
+	gdb --args $(SIM_BUILD)/$(TOPLEVEL) $(PLUSARGS)
```
Comment (imphil) [inline cocotb/share/makefiles/simulators/Makefile.verilator]: as above.

File: cocotb/share/makefiles/simulators/Makefile.verilator (hunk 1 above)
Comment (eric-wieser) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: > can you indent them with one more tab?

Note that doing this results in an extra tab being passed to the shell - makefiles supposedly strip exactly one leading tab from wrapped lines.

Not that this matters.

Comment (markmelvin) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: > * What's the motivation for switching away from `$<`? It should be identical?

It should be, I agree.  But on my system I found that the build always failed as `$<` always expanded to just `sim_build/` and the `$(TOPLEVEL)` suffix was always lost.  I confirmed this var was indeed set.  The only way I could make it work was to remove `$<` and specify `$(SIM_BUILD)/$(TOPLEVEL)` explicitly (which seems to be what the other makefiles do).

Comment (eric-wieser) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: Can you add `echo "$(SIM_BUILD)/$(TOPLEVEL)"` within the target, and show us the output?

Comment (markmelvin) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: Build output (trimmed for anonymity):
```
[cocotb] g++    verilator.o verilated.o verilated_dpi.o verilated_vpi.o Vtop__ALL.a   -L /home/mark/_dev/<xxx>/build/libs/x86_64 -lvpi -lgpi -lcocotb -lgpilog -lcocotbutils  -o k1_ctrl -lm -lstdc++
[cocotb] make[2]: Leaving directory '/home/mark/_dev/<xxx>/sim_build'
[cocotb] echo "sim_build/<mytoplevel>"
[cocotb] sim_build/<mytoplevel>
[cocotb] PYTHONPATH=<my_long_pythonpath>: LD_LIBRARY_PATH=/home/mark/_dev/<xxx>/build/libs/x86_64 LD_LIBRARY_PATH=/home/mark/_dev/<xxx>/build/libs/x86_64::/usr/lib:/usr/lib:/usr/lib:/usr/lib MODULE=<mymodulename> \
[cocotb] TESTCASE= TOPLEVEL=<mytoplevel> TOPLEVEL_LANG= COCOTB_SIM=1 \
[cocotb] sim_build/
[cocotb] /bin/sh: 1: sim_build/: Permission denied
[cocotb] make[1]: *** [/home/mark/_dev/<yyy>/cocotb/cocotb/share/makefiles/simulators/Makefile.verilator:60: /home/mark/_dev/<yyy>/results.xml] Error 126
[cocotb] make[1]: Leaving directory '/home/mark/_dev/<xxx>'
[cocotb] make: *** [/home/mark/_dev/<yyy>/cocotb/cocotb/share/makefiles/Makefile.sim:80: sim] Error 2
[cocotb] make: Leaving directory '/home/mark/_dev/<xxx>'
```

Comment (eric-wieser) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: And to be clear, `<mytoplevel>` contains no whitespace?

Comment (markmelvin) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: There is an executable in `sim_build/` with run permissions called `<mytoplevel>`.  Replacing `$<` with `$(SIM_BUILD)/$(TOPLEVEL)` executes just fine.

Comment (markmelvin) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: > And to be clear, `<mytoplevel>` contains no whitespace?

No, no whitespace. Just an underscore in the middle.

Comment (eric-wieser) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: I'm asking whether the name `<mytoplevel>` contains any characters other than `a-zA-z0-9`

Comment (eric-wieser) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: Can you add `echo "[$<]"` within the recipe too?

Comment (markmelvin) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: I suppose my toplevel name is not so top-secret.  It is `k1_ctrl`.  Here is the output of both echo statements:

```
[cocotb] echo "sim_build/k1_ctrl"
[cocotb] sim_build/k1_ctrl
[cocotb] echo "[sim_build/]"
[cocotb] [sim_build/]
```

Comment (eric-wieser) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: Something deeply weird seems to be happening here. Does something like this still fail?
```
bad_target: $(SIM_BUILD)/$(TOPLEVEL)
    echo "[$<]"
```
```make bad_target```

Comment (markmelvin) [reply cocotb/share/makefiles/simulators/Makefile.verilator]: Interesting.  That works.  And running make where I define TOPLEVEL as an environment variable on the command line works fine with `Makefile.verilator` as well.

I have a different way of executing things where I spawn `cocotb` in a subprocess from a pytest-based runner.  I guess it has something to do with that machinery?

**Comment by ktbarrett** [comment]
@markmelvin Did you resolve the issue with the Makefile? Are you still interested in getting this change in? I'll close in a week if there is no response.

**Comment by ktbarrett** [comment]
No activity, no response. Reopen and rebase if you are interested in getting this done.

**Pull request closed without merging**